Namespaced actors (#4192)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/actors/namespaced-actors.md

---
type: docs
title: "Namespaced actors"
linkTitle: "Namespaced actors"
weight: 40
description: "Learn about namespaced actors"
---

> Namespaced actors use the multi-tenant Placement service. With this service, in a scenario where each tenant has its own namespace, sidecars belonging to a tenant named "Tenant A" won't receive placement information for "Tenant B". 

Reviewer: Update to the following put then place this at end of this article. This is an implementation detail and not that important. 

```suggestion
Namespaced actors use the multi-tenant Placement service. With this control plane service where each application deployment has its own namespace, sidecars belonging to an application in namespace "ActorA" won't receive placement information for an application in namespace "ActorB". 
```

Namespacing in Dapr provides isolation, and thus multi-tenancy. With namespacing, the same actor type can be deployed into different namespaces. Instances of actors can call across these namespaces. 

{{% alert title="Note" color="primary" %}}
Note that the namespace is not part of the actor name (that is an actor id), namespace is used for the isolation of the deployed actor services.
{{% /alert %}}

> 

Reviewer: {{% alert title="Note" color="primary" %}}
Note that each namespaced actor deployment must use its own separate state store, especially if the same actor type is used across namespaces. In other words, no namespace information is written as part of the actor record, and hence separate state stores are required for each namespace. See "Configuring actor state stores for namespacing" section below.
{{% /alert %}}

Put the proper link in for the "Configuring actor state stores for namespacing"

## Creating and configuring namespaces

You can use namespaces either in self-hosted mode or on Kubernetes.

{{< tabs "Self-Hosted" "Kubernetes">}}

{{% codetab %}}
In self-hosted mode, you can specify the namespace for a Dapr instance by setting [the `NAMESPACE` environment variable]({{< ref environment.md >}}).

{{% /codetab %}}

{{% codetab %}}
On Kubernetes, you can create and configure namepaces when deploying actor applications. For example, start with the following `kubectl` commands:

```bash
kubectl create namespace namespace-actorA
kubectl config set-context --current --namespace=namespace-actorA
```

Then, deploy your actor applications into this namespace (in the example, `namespace-actorA`).

{{% /codetab %}}

{{< /tabs >}}

## Configuring actor state stores for namespacing

> While you could use different physical databases for each actor namespace, some state store components provide a way to logically separate data by table, prefix, collection, and more. This allows you to use the same physical database for multiple namespaces, as long as you provide the logical separation in the Dapr component definition.

Reviewer: ```suggestion
Each namespaced actor deployment **must** use its own separate state store. While you could use different physical databases for each actor namespace, some state store components provide a way to logically separate data by table, prefix, collection, and more. This allows you to use the same physical database for multiple namespaces, as long as you provide the logical separation in the Dapr component definition.

Some examples are provided below.
```

### Example 1: By a prefix in etcd

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: statestore
spec:
  type: state.etcd
  version: v2
  metadata:
  - name: endpoints
    value: localhost:2379
  - name: keyPrefixPath

>     value: namespace1

Reviewer: ```suggestion
    value: namespace-actorA
```

  - name: actorStateStore
    value: "true"
```

### Example 2: By table name in SQLite

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: statestore
spec:
  type: state.sqlite
  version: v1
  metadata:
  - name: connectionString
    value: "data.db"
  - name: tableName

>     value: "namespace1"

Reviewer: ```suggestion
    value: "namespace-actorA"
```

  - name: actorStateStore
    value: "true"
```

### Example 3: By logical database number in Redis

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: statestore
spec:
  type: state.redis
  version: v1
  metadata:
  - name: redisHost
    value: localhost:6379

>   - name: redisPassword

Reviewer: We should show best practices for using password everywhere, like this https://docs.dapr.io/operations/components/component-secrets/

```
  - name: redisPassword
    secretKeyRef:
      name: redis-secret
      key:  redis-password
  - name: actorStateStore
    value: "true"
  - name: redisDB
    value: "1"
auth:
  secretStore: <SECRET_STORE_NAME>
```

---
type: docs
title: "Namespaced actors"
linkTitle: "Namespaced actors"
weight: 40
description: "Learn about namespaced actors"
---

> Namespacing in Dapr provides isolation, and thus multi-tenancy. With namespacing, the same actor type can be deployed into different namespaces. Instances of actors can call across these namespaces. 

Reviewer 2: Same comment here, we can only call actors in the same namespace.

---
type: docs
title: "Namespaced actors"
linkTitle: "Namespaced actors"
weight: 40

> description: "Learn about the multi-tenant placement service in actors"

Reviewer: ```suggestion
description: "Learn about namespaced actors"
```

---

> Namespaced actors use the multi-tenant placement service. With this service, in a scenario where each tenant has its own namespace, sidecars belonging to a tenant named "Tenant A" won't receive placement information for "Tenant B". 

Reviewer: Namespacing in Dapr provides isolation and hence multi-tenancy. Dapr supports namespaced actors. This means the same actor type can be deployed into different namespaces and instances of actors can call across the namespaces. Note that the namespace is not part of the actor name (that is an actor id), namespace is used for the isolation of the deployed actor services. 

On Kubernetes you use namespaces when deploying the actor applications. For example

kubectl create namespace namespace-actorA
kubectl config set-context --current --namespace=namespace-actorA

Then deploy your actor applications into this namespace.

In self hosted mode you can specify the namespace for a Dapr instance by setting the NAMESPACE environment variable.

---
type: docs
title: "Namespaced actors"
linkTitle: "Namespaced actors"
weight: 40
description: "Learn about the multi-tenant placement service in actors"
---

Namespaced actors use the multi-tenant placement service. For example, sidecars belonging to "Tenant A" don't receive placement information meant for "Tenant B". 

<!-- need diagram --> 

In order for multiple tenants to have actor types and/or IDs with the same name, every namespace should have its own state store. Otherwise, apps in different namespaces with the same actor type and/or ID may overwrite each other's data in the state store.

## Example

In the following example...

<!-- need example? ok to use from placement API? --> 

## Migrating data to a new namespace

If you're moving to a new namespace and starting to use a new state store, make sure you migrate your data.

## Backwards compatibilty

> Namespaced actors are backwards compatible, allowing you to block newer sidecar versions from seeing the actor types within older sidecar versions.

Reviewer 2: Namespaced actors are backwards compatible for deployments that use mTLS,  because the sidecar's namespace is inferred from the Spiffe ID, allowing for multi-tenancy out of the box.
When mTLS is not enabled we default to what is explained below.

Depending on whether mTLS is enabled, the namespace is either:
- Verified through Spiffe ID (mTLS enabled), or
- Accepted as-is (mTLS not enabled)

### With mTLS enabled

Let's say you've enabled mTLS. As soon as the placement server is updated, the sidecars in namespace X (A, B, and C) can see each other’s actor types, and no others. The same is true for sidecars D, E, and F in namespace Y.

> <img src="/images/namespaced-actors-with-mtls.png" width=900>

Reviewer 2: Can we use pre 1.14 and 1.14+ instead of new and old version?
